### ISSUE TITLE
Multivariate t-tests not working

### ISSUE THREAD

**Issue by ktavabi** [ktavabi/badbaby#21]
@pettetmw this is more along the lines of what I have in mind for TFR script. That being said I don't think [this implementation is correct](https://github.com/ktavabi/badbaby/blob/747d23a60cbe46da1551deef7797ffb8d73a7e10/badbaby/analysis/write_assr_tfr.py#L85)

@larsoner do you have any time to look at the t-tests here?

**Comment by larsoner** [comment]
What is `tcirc` supposed to be computing? It doesn't look like a Hotelling T2 at least (multivariate extension of the t test). A quick naive search suggests Watson's test can be used to test the null hypothesis that data are uniformly circularly distributed, is that the intended calculation here?

**Comment by ktavabi** [comment]
`tcirc` is a test for phase-locking using the complex transform of `itc`. Answer me this
1. For the Stockwell `AverageTFR` how do I derive the complex numbers?
2. Also, to confirm the order of operation:
  1. Compute individual subject AverageTFR (.h5)
  2. Compute individual level `pow` &  `coh` measures ### _as evoked objects_?
  3. Compute sensor level grand averaged TFR

**Comment by pettetmw** [comment]
I found the following document in a Slack posting last May.  I've also attached the original reference.
[Summary of T2circ formulation.docx](https://github.com/ktavabi/badbaby/files/7010049/Summary.of.T2circ.formulation.docx)
[VictorMast1991.pdf](https://github.com/ktavabi/badbaby/files/7010051/VictorMast1991.pdf)

**Comment by pettetmw** [comment]
> `tcirc` is a test for phase-locking using the complex transform of `itc`. Answer me this
> 
> 1. For the Stockwell `AverageTFR` how do I derive the complex numbers?
> 2. Also, to confirm the order of operation:
> 3. Compute individual subject AverageTFR (.h5)
> 4. Compute individual level `pow` &  `coh` measures ### _as evoked objects_?
> 5. Compute sensor level grand averaged TFR

1. Stockwell doesn't offer the parameter combo `average=False, output='complex'` used to get the complex `EpochsTFR`.
3. Only if you pick a single frequency (or summarize over multiple), and only if you need some method not available to you in the `TFR` classes.

**Comment by pettetmw** [comment]
> @pettetmw this is more along the lines of what I have in mind for TFR script. That being said I don't think [this implementation is correct](https://github.com/ktavabi/badbaby/blob/747d23a60cbe46da1551deef7797ffb8d73a7e10/badbaby/analysis/write_assr_tfr.py#L85)
> 
> @larsoner do you have any time to look at the t-tests here?

I'm pretty sure the code works out to the same value as Victor&Mast, but I'll double-check.  The implementation takes an algebraic shortcut by averaging the real and imaginary variances.  Assuming that checks out correct, @ktavabi should I revise the code anyway to more closely follow the Victor&Mast formulation?

**Comment by larsoner** [comment]
@pettetmw @ktavabi it looks like you've figured out the math, but let me know if you want me to look at something later. At some point I do want to properly understand the relationship between the Victor & Mast formulation, Watson's test, and Hotelling T2...

**Comment by ktavabi** [comment]
@pettetmw  If you think you can give me a verified Python module (function) that computes the Victor & Mast formulation of t-circ from a Morlet/Multitaper MNE class output, then that is the best use of the remainder of your time in this project.

**Comment by pettetmw** [comment]
On it.

**Comment by pettetmw** [comment]
Matlab verification revealed one coding error, which needs patch in mne-python version.  Unfortunately, both Hotelling and T2circ require different scaling to get the histograms of simulated T2 values to line up with their nominal F-distributions.  They are clearly both F-distributed, but the need for scale factors implies the formulations I'm using in Matlab are still somehow incorrect.  Next week, I'll try comparing output of mne-python and Matlab versions when given same input to determine if they are at least generating the same output.

**Comment by ktavabi** [comment]
@pettetmw you should break till we have input from @larsoner to move forward with comparisons between the test types. I am making headway with ASSR data on my end and have most of what I need. 

<img width="352" alt="tone" src="https://user-images.githubusercontent.com/4381290/131209346-ee524b28-c0ee-43a1-920c-193c71210f9e.png">
That's textbook phase locking!

Feel free to see what I am doing in [write-up](https://github.com/ktavabi/badbaby/tree/write-up) branch, I am hoping Eric will have time to review on his stack. There seems to be too much uncertainty with your implementation of the `t-circ` to be of immediate use.

**Comment by larsoner** [comment]
> Unfortunately, both Hotelling and T2circ require different scaling to get the histograms of simulated T2 values to line up with their nominal F-distributions.

I read a little bit of the Victor Mast and they argue that t2circ is better than the other measures I mentioned for ASSR (and searching several more recent reviews mention/confirm it has higher power than a Hotelling T2 for example). I've read the PDF and also [this powerpoint](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.78.4830&rep=rep1&type=pdf) (which has a slightly different, seemingly not equivalent formulation) and neither seems to match when I create normally distributed, zero-mean "Fourier coefficients", calculate t2_circ, and compare the histogram of `M * t2_circ` (blue) to `F(2, 2 * M - 2)` distribution (orange), which it seems like should match, but they're not particularly close:

![Screenshot from 2021-08-30 17-59-59](https://user-images.githubusercontent.com/2365790/131411449-d5c42be5-89b6-46ab-a544-7bf4b92c9a35.png)

@pettetmw if you want to work off of the same script as me, this is what I had in ~30 lines:

https://gist.github.com/larsoner/8868785b4909ea4b72b4274c133522e9

**Comment by pettetmw** [comment]
@larsoner , I get the same thing.  If I apply scale factors to both T2 and F, I can get them to line up:
![image](https://user-images.githubusercontent.com/9087138/131540058-6254ce15-4c1d-4d9e-8d84-9a1202285434.png)

**Comment by pettetmw** [comment]
The scale factors depend on the number of trials... in this example, 30.  This also for your consideration:
![image](https://user-images.githubusercontent.com/9087138/131540255-2b982f45-cc89-44d0-a563-b73585548367.png)

**Comment by pettetmw** [comment]
The difference in denominator degrees of freedom does apparently give T2_circ a sensitivity advantage over Hotelling's for trial counts down below 15 or so.  My concern, given our troubles verifying the formulations, is that one man's sensitivity is another man's bias.

**Comment by pettetmw** [comment]
FWIW, here's what I get using Wikipedia's formulation of Hotelling's.  (It still requires the vertical scale factor on F, but I figure that's my own ignorance about how to properly normalize a monte-carlo histogram.)
![image](https://user-images.githubusercontent.com/9087138/131542850-c653cb13-388b-4d15-921c-62175addee64.png)

**Comment by larsoner** [comment]
So Hoteling works as it should, but t2_circ does not? I'd be inclined just to use Hotelling in that case, knowing that we've lost a little bit of sensitivity. If the effect is robust it might be good enough.

> (It still requires the vertical scale factor on F, but I figure that's my own ignorance about how to properly normalize a monte-carlo histogram.)

I spent at least 20 minutes trying to remember/figure out the scaling that works to get from the `hist` to PDF-like values and I couldn't do it, either :) But really it's the shape that matters here, and it does look like the Hotelling matches up. I could also get the t2_circ values to match up, but I couldn't find a consistent factor that depended on L or L-1 or some power of those (or their reciprocals), which isn't good.

TL;DR: t2_circ is a pain, I'd use Hotelling T2 for now and if sensitivity seems like it's a problem then come back to t2_circ later

**Comment by larsoner** [comment]
FWIW this is what I coded for Hotelling T2 some time ago:

https://github.com/LABSN/mnefun/blob/c5a1b5b15e4c2b338ef6de14323b576d67b85f53/mnefun/_stats.py#L86

Feel free to check to see if your code matches what it produces! It would be a nice sanity check on my code

**Comment by larsoner** [comment]
... though it makes lots of assumptions like there are 3 dimensions (for X/Y/Z in free orientation), so it'll actually probably be a pain. So maybe just worth looking at to see if the computations otherwise match.

**Comment by pettetmw** [comment]
Seriously, you started from Wikipedia too?

**Comment by larsoner** [comment]
Do you mean SPSS? I can't remember how I validated it, I don't think I used SPSS but probably some other implementation(s). Validating with random data like in the gist is also good. I was hoping for example that with `p < 0.05` and "Fourier components" for thousands of "sensors" generated using `rng.randn` as in the gist I would end up with ~20% being significant for example...

**Comment by pettetmw** [comment]
> Do you mean SPSS?

No, there's two PDF links in block quotes, `_stats.py`, line 93.

>I was hoping for example that with `p < 0.05` and "Fourier components" for thousands of "sensors" generated using `rng.randn` as in the gist I would end up with ~1/20 = ~5% being significant for example...

That's roughly how I have it implemented in Matlab as well.

**Comment by pettetmw** [comment]
@ktavabi , other than satisfying my own curiosity about @larsoner's `Hotelling` implementation and tests, I don't think we need further evaluation:  `Hotelling` is the safer bet, and `t2_circ `is not ready for prime time.  While I'd certainly enjoy debugging an algorithm from a 30-year-old methods publication, I don't think we should hold up the `badbaby` show to indulge my enthusiasms.

**Comment by pettetmw** [comment]
@ktavabi, do you think it's a worthwhile sanity check for me to try and get your `badbaby/analysis/assr_tfr_analysis.ipynb` to work on my local copy of `badbaby` dataset?

**Comment by ktavabi** [comment]
> one man's sensitivity is another man's bias.
Is that the same as one man's variance is another man's bias?



> do you think it's a worthwhile sanity check for me to try and get your `badbaby/analysis/assr_tfr_analysis.ipynb` to work on my local copy of `badbaby` dataset?

Yes, be great if either of you could get the notebook to run on another stack. I still haven't got to testing 🙊but when I do Hotelling it is.

**Comment by pettetmw** [comment]
Hi @larsoner, for lack of a better place to post this, I've had occasion to use Hotelling's T2 to test for evoked activation in a sequence of consecutive time intervals.  I installed https://dionresearch.github.io/hotelling/readme.html and ran the following test:

```
from scipy.stats import norm
from hotelling import stats
import numpy as np

n = 10000
pvals = np.zeros(10000)

for i in range(n) :
	x = norm.rvs(size=[30,5]) # normal random variates
	nhyp = np.array((0,0,0,0,0)) # null hypothesis
	p = stats.hotelling_t2( x, nhyp )
	pvals[i] = p[2]

crit05 = np.nonzero(pvals<=0.05)
crit05 = crit05[0].size / n
print( crit05 )
```
Running the above script returns values roughly consistent with nominal 5% critical value.  So, this hotelling implementation seems to be unbiased for inferential purposes.  I haven't yet checked whether T2 histogram fits to the F-distribution without scaling (e.g. Aug 31 post above).  In your opinion, should I bother? given that the shape is what matters for critical value?

**Comment by larsoner** [comment]
I would cross-check three implementations:

1. `hotelling`: https://dionresearch.github.io/hotelling/readme.html
2. `mnefun`: https://github.com/LABSN/mnefun/blob/d9a3b020731949147aed71bfee3d4f67f880cccc/mnefun/_stats.py#L86
3. @drammock's implementation (location TBD)

If they all agree I think we're safe to use any of them. If they don't agree then we should figure out why.

FWIW the `mnefun` one is only adapted to the case where there are 3 variates, so that's the case I would test. If it's not obvious what to do I could quickly write a little script to get started

**Comment by drammock** [comment]
> @drammock's implementation (location TBD)

I didn't end up implementing Hotelling's T^2 for the F2F connectivity because it was a within-subjects analysis (and so ends up being a 1-sample test).

https://github.com/ilabsbrainteam/face-to-face-connectivity/blob/3c4a4590b382b8b7e94b7c19aa1d66d309353f39/analysis/graph-level-stats.py#L120-L124

**Comment by ktavabi** [comment]
It looks to me the mnefun implementation of `hotelling` is for the express purpose of dealing with source orientations in the `.stc` from vector source estimates. So it mks sense to see whether a `logit` classifier on source orientations can preserve the major `ori` for a given source estimate in a model. I.e., yield [this matrix as the target](https://github.com/LABSN/mnefun/blob/d9a3b020731949147aed71bfee3d4f67f880cccc/mnefun/_stats.py#L121)

According to the docs, Hotelling is a generalization of the Student's ratio, a multivariate test with f-distribution and thus more robust than the favorited 1-sample univariate test case that exists only in labs.

**Comment by drammock** [comment]
> a multivariate test with f-distribution and thus more robust than the favorited 1-sample univariate test case that exists only in labs

I find this remark a bit confusing, there are one-sample and two-sample versions of Hotelling's T^2, and I'm not clear where robustness comes in. Perhaps what I said above was misleading? When I said

> I didn't end up implementing Hotelling's T^2

what I meant was "I only implemented the one-sample (chi-squared distributed) version of Hotelling's test, I didn't do a *general* implementation that can handle two sample tests".
